### PR TITLE
PE-1439: Date calculation

### PR DESCRIPTION
**Pull request by matibat**

### CODE REVIEW

File: src/common.ts [hunk 1]
```diff
@@ -100,3 +100,63 @@ export async function validateTxTip(node: GQLNodeInterface, ardriveOracle: ArDri
...
+export function weeksDiffInEST(prev: Date, curr: Date): number {
+	const prevEstDate = dateToEST(prev);
+	const currEstDate = dateToEST(curr);
+
+	let weeksCount = 0;
+	const cursorDate = toSunday(prevEstDate);
+
+	while (currEstDate.getTime() > cursorDate.getTime()) {
+		const cursorDay = cursorDate.getDate();
+		const daysPerWeek = 7;
```
Comment (karlprieb) [inline src/common.ts:133]: sunday is 0 and saturday is 6

File: src/common.ts [hunk 2]
```diff
@@ -100,3 +100,63 @@ export async function validateTxTip(node: GQLNodeInterface, ardriveOracle: ArDri
...
+		cursorDate.setDate(cursorDay + daysPerWeek);
+		if (cursorDate.getTime() < currEstDate.getTime()) {
+			weeksCount++;
+		}
+	}
+	return weeksCount;
+}
+
+function toSunday(date: Date) {
+	const day = date.getDay() || 7;
```
Comment (karlprieb) [inline src/common.ts:143]: You need to use 0 if you want to target sunday
```suggestion
	const day = date.getDay() || 0;
```

File: src/common.ts [hunk 3]
```diff
@@ -100,3 +100,63 @@ export async function validateTxTip(node: GQLNodeInterface, ardriveOracle: ArDri
...
+	const newDate = new Date(date.getTime());
+	if (day !== 7) {
```
Comment (karlprieb) [inline src/common.ts:145]: You need to use 0 if you want to target sunday
```suggestion
	if (day !== 0) {
```

File: src/common.ts (hunk 1 above)
Comment (matibat) [reply src/common.ts:133]: What I do in the line below is to sum 7 days to the current one

Comment (karlprieb) [reply src/common.ts:133]: Ohhh you're right

File: src/common.ts [hunk 4]
```diff
@@ -100,3 +100,63 @@ export async function validateTxTip(node: GQLNodeInterface, ardriveOracle: ArDri
...
+		newDate.setHours(-24 * (day - 1));
+	}
+	newDate.setHours(0);
+	newDate.setMinutes(0);
+	newDate.setSeconds(0);
+	return newDate;
+}
```
Comment (karlprieb) [inline src/common.ts]:
```suggestion
function toSunday(date: Date) {
	const sunday = 0;
	const day = date.getDay();
	const newDate = new Date(date.getTime());

	if (day !== sunday) {
		while (newDate.getDay() > sunday) {
			newDate.setDate(newDate.getDate() - 1);
		}
	}
	newDate.setHours(0);
	newDate.setMinutes(0);
	newDate.setSeconds(0);
	return newDate;
}
```
This change should return the last Sunday :)

**Review by karlprieb — CHANGES_REQUESTED** [review]
`weeksDiff` is not returning the right value. It will after the changed that I suggested.
While it returns the right value if I test with transactions from last week the wallet that did these transactions will not have `lastWeek` populated, just `total`.

File: src/daily_output.ts [hunk 5]
```diff
@@ -240,10 +247,14 @@ export class DailyOutput {
 		const queryDate = new Date(queryTimestamp);
 		const previousTimestamp = this.latestTimestamp * 1000;
 		const previousDate = new Date(previousTimestamp);
```
Comment (karlprieb) [inline src/daily_output.ts:249]: `previousDate` should be the date from the last transaction that we processed and not the last time that we ran the script.

File: src/common.ts (hunk 4 above)
Comment (matibat) [reply src/common.ts]: I simplified that a little bit: 8f3d45f6e969a73718335566c6b4cf9d5a483335
And added tests just to be sure: 53e6f06bb8af00406f4af4979ffd178c528908c6

File: src/common.test.ts [hunk 6]
```diff
@@ -55,4 +55,72 @@ describe('common methods', () => {
...
+
+			expect(diffInMilliseconds).to.be.lessThan(aWeekInMilliseconds);
+			expect(dateOnPrevSunday.getDay()).to.equal(0);
+			expect(dateOnPrevSunday.getHours()).to.equal(0);
+			expect(dateOnPrevSunday.getMinutes()).to.equal(0);
+			expect(dateOnPrevSunday.getSeconds()).to.equal(0);
+			expect(dateOnPrevSunday.getMilliseconds()).to.equal(0);
+		});
+
+		it('returns a date in the same day if already on sunday', () => {
```
Comment (fedellen) [inline src/common.test.ts:111]: Getting this test failure in my env:

```shell
  1) common methods
       dateToSunday method
         returns a date in the same day if already on sunday:

      AssertionError: expected 597600003 to be at most 86400000
      + expected - actual

      -597600003
      +86400000
```

Comment (matibat) [reply src/common.test.ts:111]: We are now using a specific timezone to be deterministic with the expectations
17920ab7d22fd3f7b75e5d83e27d1e80ccb9acec

Comment (matibat) [reply src/common.test.ts:111]: Nice catch! :sunglasses:

File: src/common.ts (hunk 4 above)
Comment (karlprieb) [reply src/common.ts]: Clever 👍

**Review by karlprieb — APPROVED** [review]

**Review by fedellen — APPROVED** [review]
LGTM

**Pull request merged**